### DISCUSSION TITLE
How to make changes to an environment installed from a repo?

### DISCUSSION THREAD

**Discussion by kiranikram** [community/community#22126] [New to GitHub]
<p>Hi, I need to utilise code that I am allowed to install from a particular repo. What are the correct steps to install it, make the necessary changes + additions, and then push it to my own github account ?<br>
Thanks!</p>

**Comment by tajmone** [comment]
<p>If I’ve understood correctly, you’re asking about GitHub’s “Fork and Pull Request” model, which is the workflow adopted by GitHub to collaborate to repositories created by others.</p>
<p>You might find a good introductory article at this link:</p>
<aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://reflectoring.io/assets/img/logo/favicon.ico" class="site-icon" width="15" height="16">
      <a href="https://reflectoring.io/github-fork-and-pull/" target="_blank" rel="nofollow noopener" title="12:00AM - 02 January 2017">reflectoring.io – 2 Jan 17</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/361;"><img src="https://user-images.githubusercontent.com/11064292/181068969-51a0d391-72a5-46fb-8663-54054bc4aa60.jpeg" class="thumbnail" width="690" height="361" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/b/b921f7716b0e07933b411f8299f329c3d259f196_2_690x361.jpeg, https://aws1.discourse-cdn.com/github/optimized/2X/b/b921f7716b0e07933b411f8299f329c3d259f196_2_1035x541.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/b/b921f7716b0e07933b411f8299f329c3d259f196.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b921f7716b0e07933b411f8299f329c3d259f196_2_10x10.png"></div>

<h3><a href="https://reflectoring.io/github-fork-and-pull/" target="_blank" rel="nofollow noopener">Github's Fork &amp; Pull Workflow for Git Beginners</a></h3>

<p>If you are new to Git and/or GitHub, it's easy to get overwhelmed by the different workflow models you can use. This post explains the basic fork and pull workflow model that is used on a lot of GitHub repositories.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>The article will provide you with a bird-eye view of the whole process, for the practical details of all the operations involved you’ll find topic-dedicated tutorials on GitHub Help:</p>
<p><a href="https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/github-flow" class="onebox" target="_blank" rel="nofollow noopener">https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/github-flow</a><br>
<a href="https://guides.github.com/introduction/flow/" class="onebox" target="_blank" rel="nofollow noopener">https://guides.github.com/introduction/flow/</a></p>
<p>If you’re new to Git and GitHub, you might consider working with GitHub Desktop, the free and open source GUI Git client based on the GitHub Workflow:</p>
<aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/pinned-octocat.svg" class="site-icon" width="36" height="36">
      <a href="https://desktop.github.com/" target="_blank" rel="nofollow noopener">GitHub Desktop</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="" height="">

<h3><a href="https://desktop.github.com/" target="_blank" rel="nofollow noopener">GitHub Desktop</a></h3>

<p>Simple collaboration from your desktop</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>This tool will simplify the most common Git operations involved in code collaboration; and thanks to the graphic interface, it might also help gain better insight into repositories’s structures.</p>

**Comment by kiranikram** [comment]
<p>Thanks! So if I have forked and cloned it eg to say Pycharm, and I want to make additions on the original code as a project of my own, do I need to a)create another repo on Github to which I push my changes b)Am I supposed to be working on a ‘branch’ of the original?<br>
To clarify, once I get the code on which to build from the original repo, I dont need to utilize/contribute anything to the original authors; its now a project of my own.<br>
previously when I have forked projects, it assumes I want to contribute to the original and it gives me an error, so I want to do it correctly this time; Thank you :slight_smile:</p>

**Comment by tajmone** [comment]
<aside class="quote no-group" data-username="kiranikram" data-post="3" data-topic="120878">
<div class="title">
<div class="quote-controls"></div>
 kiranikram:</div>
<blockquote>
<p>do I need to a)create another repo on Github to which I push my changes</p>
</blockquote>
</aside>
<p>We’ll yes, but you’ve already done that when you forked the repository. Your fork is a <em>copy</em> of the <em>upstream</em> (i.e. original) repository, and to make a contribution (a <em>pull request</em>) you need to first push the changes to your fork, then from GitHub’s home page of your fork you can create the pull request to the <em>upstream</em> project.</p>
<aside class="quote no-group" data-username="kiranikram" data-post="3" data-topic="120878">
<div class="title">
<div class="quote-controls"></div>
 kiranikram:</div>
<blockquote>
<p>b)Am I supposed to be working on a ‘branch’ of the original?</p>
</blockquote>
</aside>
<p>Yes, this is very important! You should not apply (i.e. <em>commit</em>) changes directly to the <code>master</code> branch of your fork — because your fork depends on the upstream repo, you need to keep your <code>master</code> branch in synch with the <code>master</code> branch of the upstream, otherwise the two codebases will diverge and you won’t be able to integrate further updates from the upstream repo into your fork.</p>
<p>Also, some repositories will only accept pull requests on specific branches (e.g. a <code>devel</code> branch, or some other name). Even in this case, you must keep that branch clean (i.e. use it only to pull changes from the upstream) and apply your changes to a branch created from the branch which you’ll be creating the pull request to (often <code>master</code> branch, but not always).</p>

**Comment by tajmone** [comment, marked as answer]
<aside class="quote no-group" data-username="kiranikram" data-post="3" data-topic="120878">
<div class="title">
<div class="quote-controls"></div>
 kiranikram:</div>
<blockquote>
<p>To clarify, once I get the code on which to build from the original repo, I dont need to utilize/contribute anything to the original authors; its now a project of my own.</p>
</blockquote>
</aside>
<p>This depends on the license terms really. Most open source licenses allow you to do this, as long as you   honor the license terms. But some projects might demand that you contribute changes to the upstream.</p>
<aside class="quote no-group" data-username="kiranikram" data-post="3" data-topic="120878">
<div class="title">
<div class="quote-controls"></div>
 kiranikram:</div>
<blockquote>
<p>previously when I have forked projects, it assumes I want to contribute to the original and it gives me an error, so I want to do it correctly this time;</p>
</blockquote>
</aside>
<p>Most likely, you cloned locally the upstream repository directly, so when you tried to push your changes you were presented with an “access denied” error, for you don’t have write permission on the original repository. Which is why you need to first fork the repository, so you have your own copy to which you (and only you, initially) have write access.</p>
<p>Understand that pull requests are natively part of Git, is an added layer by services like GitHub to allow easier cross-collaboration with people who don’t know each other. I.e. a way to allow contributions also from those who don’t have write access to the project.</p>

**Comment by tajmone** [comment]
<p>As a last note, beware that when you fork a repository to you only get the <code>origin</code> <em>remote</em>, which is a pointer back to your forked repository. But in order to integrate updates from the original project you’ll also need to manually add the <code>upstream</code> remote, pointing to the original repository — neither Git nor GitHub do this automatically for you, you’ll have to do it manually.</p>
<p>Basically, when you start working on your changes you are doing so from a branch which you created from a given snapshot of the repository. In the meantime (let’s say you work on your feature for 5 days) the upstream repository might have been updated, so your branch is behind that of the original repo.</p>
<p>You’ll need to set the <code>upstream</code> remote in order to pull updates from the upstream into your local clone (of your fork) and then push the updates to your fork on GitHub, as well as to rebase your contribution branch, so that your PR will not be out of synch with the destination branch when you create the PR.</p>
<p>It might sound all a bit complicated, but it’s actually simpler than it sounds, and it will all click-in once you actually make a PR or need to integrate updates from the upstream. Just keep in mind that Git is a <em>distributed</em> version control system, so you’re free to <em>pull</em> (i.e. get) updates from multiple versions of the same project (the original upstream repo, your fork, other forks, etc.). Branches are what allow you to keep different versions of the same code base (but with common ancestry)  in a single repository.</p>
<p><a href="https://docs.github.com/en/github/using-git/managing-remote-repositories" class="onebox" target="_blank" rel="nofollow noopener">https://docs.github.com/en/github/using-git/managing-remote-repositories</a><br>
<a href="https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes" class="onebox" target="_blank" rel="nofollow noopener">https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes</a></p>

**Comment by kiranikram** [comment]
<p>Thank you, your answers have been invaluable! I appreciate the help!</p>
